lasagne/layers/pool.py

```diff
@@ -291,6 +292,55 @@ def __init__(self, incoming, pool_size, stride=None,
 # TODO: add MaxPool3DLayer
 
 
+class Upscale2DLayer(Layer):
+    """
+    2D upscaling layer layer
+
+    Performs 2D upscaling over the two trailing axes of a 4D input tensor.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or tuple
+        The layer feeding into this layer, or the expected input shape.
+
+    scale_factor : integer or iterable
+        The scale factor in each dimension. If an integer, it is promoted to
+        a square scale factor region. If an iterable, it should have two
+        elements.
+
+    **kwargs
+        Any additional keyword arguments are passed to the :class:`Layer`
+        superclass.
+    """
+
+    def __init__(self, incoming, scale_factor, **kwargs):
+        super(Upscale2DLayer, self).__init__(incoming, **kwargs)
+
+        self.scale_factor = as_tuple(scale_factor, 2)
+
+        if self.scale_factor[0] < 1 or self.scale_factor[1] < 1:
+            raise ValueError('Scale factor must be >= 1, not {0}'.format(
+                self.scale_factor))
+
+    def get_output_shape_for(self, input_shape):
+        output_shape = list(input_shape)  # copy / convert to mutable list
+        if output_shape[2] is not None:
+            output_shape[2] *= self.scale_factor[0]
+        if output_shape[3] is not None:
+            output_shape[3] *= self.scale_factor[1]
+        return tuple(output_shape)
+
+    def get_output_for(self, input, **kwargs):
+        y = self.scale_factor[0]
+        x = self.scale_factor[1]
+        upscaled = input
+        if x > 1:
+            upscaled = T.extra_ops.repeat(upscaled, x, 3)
+        if y > 1:
+            upscaled = T.extra_ops.repeat(upscaled, y, 2)
```

Comment:
That's a bit confusing. At first glance, it looks like the two scale factors are applied to the two dimensions in the wrong order. Why not:
```python
    x, y = self.scale_factor
    upscaled = input
    if x > 1:
        upscaled = T.extra_ops.repeat(upscaled, x, 2)
    if y > 1:
        upscaled = T.extra_ops.repeat(upscaled, y, 3)
```
If you're worried about a discrepancy with image height and width, call the factors `f1` and `f2` or `a` and `b`.